### ISSUE TITLE
Accessibility support for linux needs to be implemented

### ISSUE THREAD

**Issue by webczat** [AvaloniaUI/Avalonia#14275]
# introduction

As a blind user which daily uses linux, I am unable to use any avalonia based applications because they are inaccessible to screenreaders, while such support exists for windows and mac. I would like accessibility support for linux to be added.

# Some details

I don't know how avalonia architecture looks like, so can't comment on that part, however I can share some things I gathered about the linux accessibility stack.

Because avalonia does not use any ui toolkit under the hoot and is based directly on X (and possibly wayland, doesn't really matter that much), there are two options to implement accessibility:
1. Use ATK library as a dependency and implement ATK interfaces.
2. Expose application directly to at-spi, the actual linux accessibility stack.
Option 2 is recommended for cases like avalonia, even gtk4 stopped using ATK in it's implementation, even though GTK was atk's main user. However, below I will try to describe both options, especially that option1 is probably better documented.

## Option 1: use ATK

Basically, that requires adding ATK as a dependency. In this case, for each avalonia control, you need a peer object implementing accessibility. This peer object is a GObject-based object inheriting from AtkObject and implementing one or more atk interfaces as appropriate for the object. The avalonia control itself doesn't need to be a GObject object, but one instance of a control needs to be bound to one instance of the AtkObject subclass appropriate for it. This probably can be done dynamically based on avalonia's way to indicate accessibility support for custom controls, if any.

Properties of a control need to be exposed via the atk object as appropriate, and also, any change in properties, state changes or events happening which influence the accessibility state need to be propagated through the atk peer.

ATK itself is not enough because the objects and events need to be exposed to at-spi. So, the library needs to load the at-spi2-atk bridge and initialize it first. I'm just not sure how do you integrate it with something that doesn't use the glib main loop, but you should be able to do it. I haven't done such implementation myself, I've just implemented a fake custom gtk widget with fake accessibility tree once, so my hands-on experience with that is limited.

## Option 2: using at-spi directly

So, basically, we have a special accessibility bus, on which applications are exposed, and there is at-spi registry which knows about these applications. There is a library called libatspi, but it's intended for use by other side of the equation, like screenreaders and such like to inquire applications, and it can't be used to expose them. It's done directly over dbus.

Avalonia's job would be to find the accessibility bus (instructions: https://gitlab.gnome.org/GNOME/at-spi2-core/-/blob/main/bus/README.md?ref_type=heads), to connect to it and to expose it's accessibility tree as dbus objects. The last step seems to be registering an application's root object in the registry. There are probably few other special objects to be implemented like a cache...

Generally, dbus objects for accessibility tree need to implement the Accessible interface and one or more other interfaces as appropriate. The root application object also implements the Accessible interface along with Application interface, at least from what I can tell. Introspection data for all the interfaces along with some documentation can be found here https://gitlab.gnome.org/GNOME/at-spi2-core/-/tree/main/xml?ref_type=heads but at least in one place (the event.xml file) it's said that method signatures might be incorrect, so documentation might or might not be complete, and there is lots of info missing like what's the registry's bus name or object path that you need to use when registering an app. However, an app does not need a well known bus name.

According to documentation of accessible's GetParent method and based on local dbus introspection, the root application object should be called "/org/a11y/atspi/accessible/root" with an accessible role of application. All other objects are in it's subpaths, direct descendants seem to be windows (probably with a frame role).

There is also a special object /org/a11y/atspi/cache that is to be exposed by application and that should implement the cache interface, that seems to be for getting dump of the whole accessibility tree, unsure if it's implementation is required.

Registration of an application in registryis done using the socket interface (https://gitlab.gnome.org/GNOME/at-spi2-core/-/blob/main/xml/Socket.xml?ref_type=heads). One can probably detect registry disappearing and reappearing and re-register an app if needed. Unregistering is not necessary, closing dbus connection is sufficient. Socket is implemented on /org/a11y/atspi/accessible/root object of well known bus name org.a11y.atspi.Registry, or at least that's what introspection says.

I would happily provide more info but I have never tried to implement such a thing myself without ATK. But it could be an interesting exercise which I'd prefer to do without avalonia context if at all.

# Testing

Testing of course requires playing with a screenreader like `orca`, but there is also a tool showing accessibility tree called `accerciser` which can be used to debug things. It can also show accessibility events based on filters and do other fancy things which can be helpful when debugging, but note that some errors will only appear when using an actual screenreader as how it reacts to some events is not always that obvious.

**Comment by webczat** [comment]
can you explain the last point about something not being supported by tmds.dbus? that's the most interesting one.
As for lack of docs, maybe they started to improve it after they switched their recommendations to use at-spi directly. Probably some reading sources will still be necessary.

**Comment by kekekeks** [comment]
>can you explain the last point about something not being supported by tmds.dbus?

IIRC, Tmds.DBus didn't support variants properly.

**Comment by maxkatz6** [comment]
IIRC it's supported now with Tmds.DBus.Protocol and source-generator. cc @affederaffe to confirm.

**Comment by kekekeks** [comment]
Yes, the protocol library supports manually constructed variants.

**Comment by khanshoaib3** [comment]
May I suggest using [AccessKit](https://github.com/AccessKit/accesskit). It does all the low level work so the only things to be done in avalonia would be to create and send the gui tree and updates.

**Comment by webczat** [comment]
At the point where accessibility is already implemented directly for all desktop targets? except linux it probably makes no sense. also linux support is not yet mature in access kit.

**Comment by khanshoaib3** [comment]
I think we can use AccessKit only for linux, but yeah, thinking about it, it will essentially result two systems for a11y rather than having the existing system be compatible with linux.
And yup, just looked at the docs stating an issue with Orca.

**Comment by khaledmahmoud78** [comment]
Excuse me, there are C# cross-platform frameworks compatible with Orca screen reader on Linux, such as [GTK#](https://github.com/GtkSharp/GtkSharp), [Eto.Forms](https://github.com/picoe/Eto), [AlterNET UI](https://github.com/alternetsoft/AlternetUI).
How did they solve this problem?!

**Comment by maxkatz6** [comment]
@khaledmahmoud78 they didn't solve this problem. They never had it fundamentally. At least GTK# and Eto.Forms map their controls onto GTK controls, which is a semi-native Linux framework. Just like MAUI is mapped onto mobile controls. You lose customization and flexibility, but get native controls look and build-in accessibility.

Can't say much about AlterNET.

**Comment by khaledmahmoud78** [comment]
@maxkatz6 Thanks for explanation!

**Comment by jozapl** [comment]
Does Avalonia UI already work with Orca on Linux?

**Comment by rabbitism** [comment]
> Hi @maxkatz6, thanks for proactively helping on this. We are planning an enterprise product release with Avalonia, and this may completely block BATs (aka BVTs or build verification ). As you can see, we have raised https://github.com/AvaloniaUI/Avalonia/issues/19700 asking support for this and are seeking an early fix for this if possible.
> In fact, we need a common UI Automation Framework, like how Avalonia is common for all desktops.


I would say regarding build verification or integration tests, Avalonia headless testing can cover many scenarios of traditional automation test.

**Comment by avinashkhatri1986** [comment]
> > Hi [@maxkatz6](https://github.com/maxkatz6), thanks for proactively helping on this. We are planning an enterprise product release with Avalonia, and this may completely block BATs (aka BVTs or build verification ). As you can see, we have raised [#19700](https://github.com/AvaloniaUI/Avalonia/issues/19700) asking support for this and are seeking an early fix for this if possible.
> > In fact, we need a common UI Automation Framework, like how Avalonia is common for all desktops.
> 
> I would say regarding build verification or integration tests, Avalonia headless testing can cover many scenarios of traditional automation test.

@rabbitism  can you share the framework for headless testing of Avalonia app elements and can we switch to ui to launch graphics images rendered from codecs?

**Comment by rabbitism** [comment]
> [@rabbitism](https://github.com/rabbitism) can you share the framework for headless testing of Avalonia app elements and can we switch to ui to launch graphics images rendered from codecs? consider we have created app like RDP with avalonia as framework how we validate RDP settings and all RDP screen images with headless

Unfortunately this sounds like the case that headless testing won't work.

**Comment by avinashkhatri1986** [comment]
> > [@rabbitism](https://github.com/rabbitism) can you share the framework for headless testing of Avalonia app elements and can we switch to ui to launch graphics images rendered from codecs? consider we have created app like RDP with avalonia as framework how we validate RDP settings and all RDP screen images with headless
> 
> Unfortunately this sounds like the case that headless testing won't work.

Yeah , hence we are completely blocked and waiting for immeditate resolution , if prioritized it will be great encouragement for Avlonia project to support enterprise level projects https://github.com/AvaloniaUI/Avalonia/issues/19700 is must fix now

**Comment by MikeCodesDotNET** [comment]
@avinashkhatri1986, I feel I need to clarify how open source works. Avalonia is provided as-is, without warranty, and without any obligation to implement features you demand. 

Your enterprise product's requirements, BATs, and release schedules are not our concern. 

Linux accessibility support will be implemented when:
- Someone contributes the code, or
- Someone funds the development, or 
- We get around to it in our own time 

Showing up for the first time and declaring features "must fix now" while contributing absolutely nothing is not how this works. We are not your unpaid engineering department. If this is truly blocking your enterprise product, you can: 

- Submit a PR
- Hire someone to submit a PR
- Pay us to implement it for you (our last quote for this was between €50-€70K) 
- Use a different UI framework 

It's exceptionally rude to join our community by asking us to drop everything and provide you with an "immediate resolution".  If this is critical to your business and the timeline tight, I'll be more than happy to send over a quote. Only once you've paid us will I accept this kind of attitude. 

Best regards,
Mike J

**Comment by avinashkhatri1986** [comment]
Thanks, @MikeCodesDotNET, for clarifying, and I understand your perspective. I truly respect the open-source model and the huge effort the Avalonia team has already put in. My earlier message may have come across as demanding — that wasn’t my intent, and I apologise if it sounded that way.

I’m here to be a supporter and contributor where it makes sense. I’d be happy to collaborate, contribute code, or help in other ways to make Avalonia a success. I see the value this project brings, and I’d like to be part of that journey.

**Comment by avinashkhatri-dev** [comment]
HI Team 
Thanks for the feedback! I’ve gone ahead and made some changes to extend the work previously done by [@grokys](https://github.com/AvaloniaUI/Avalonia/tree/feature/ui-automation-atspi)
.

I achieved partial success by modifying Tmds.DBus — specifically, by creating a virtual function to remove the xmlns:xsi attribute, which was being generated by an unknown layer. Using this approach, I was able to send out a custom XML format, and now the Tmds D-Bus service is listing control XML properly.

The next challenge I’m working on is AT-SPI exposure, as its structure differs from the current implementation. I’m still investigating this, but partial progress can be found here:

[Avalonia (ui-automation branch)](https://github.com/avinashkhatri-dev/Avalonia/tree/ui-automation)

[Tmds.DBus (ui-automation branch)](https://github.com/avinashkhatri-dev/Tmds.DBus/tree/ui-automation)

**Comment by avinashkhatri-dev** [comment]
More progress made now , i can see Windows and Button in pytatspi event accercisor list (but crashes in sometime), cant perform control click when done, will look on this coming week , action registration is messed up in my above branch

**Comment by grokys** [comment]
@avinashkhatri-dev before you spend any more time on this, we think that my original work was going in the wrong direction and we have the beginnings of a new implementation that we're hoping to finish in the near future.

**Comment by avinashkhatri-dev** [comment]
Nice to see the recent property additions — thanks for the continued work on this.

I’ve put my Linux investigation on hold for now since the previous path was quite broad. The newer changes appear to align better with Avalonia 11.1 and above.

If you have a recommendation, I’d appreciate guidance on which version would be best to experiment with again.

**Comment by avinashkhatri-dev** [comment]
Hi team , is this fixed with avalonia12

**Comment by timunie** [comment]
It should be much improved afaik. Would be great if you could test v12 for missing bits.

**Comment by maxkatz6** [comment]
AT-SPI2 was implemented in 12.0  - https://github.com/AvaloniaUI/Avalonia/pull/20735

**Comment by avinashkhatri-dev** [comment]
thanks for advance notice and finally supporting this , i will test and see missing pieces and also see what i can contribute here

**Comment by avinashkhatri-dev** [comment]
Good to see this improvement in xml handling — I had explored a similar direction earlier when we saw structure issues. Glad it's now aligned.
